File: lib/effs/tls_connect.js
// TLS client handshake: OpenSSL over a non-blocking fd via bun:ffi,
// with WANT_READ / WANT_WRITE parks on the underlying fd. The handle
// answers a session box {ssl, ctx, fd} (never a raw fd): send/recv/
// close unpack it. SSL_set_tlsext_host_name is a macro in OpenSSL, so
// the hostname rides SSL_ctrl (SSL_CTRL_SET_TLSEXT_HOSTNAME).
function tls_connect_raw(ip, sni, port, verify, k) {
  const sys = io_sys();
  const bad = (code, text) => ({ $: "Fail", error: io_tup(code, text) });
  const lib = tls_lib();
  if (lib === null) {
    return bad(616, "tls error");
  }
  port = Number(port);
  verify = Number(verify);
  if (!(port >= 1 && port <= 65535) || Number.isNaN(port)) {
    return bad(611, "bad input");
  }
  const at = io_addr(ip, port);
  if (at === null) {
    return bad(611, "bad input");
  }
  const fd = sys.socket(2, 1, 0);
  if (fd < 0) {
    return bad(612, "dial failed");
  }
  // The fd is non-blocking for life: SSL_connect parks on WANT_READ /
  // WANT_WRITE instead of stalling the loop.
  if (sys.fcntl(fd, 4, sys.fcntl(fd, 3, 0) | (sys.mac ? 4 : 0x800)) < 0) {
    sys.close(fd);
    return bad(612, "dial failed");
  }
  const fail = (code, text, ctx, ses) => {
    try {
      if (ses) lib.SSL_free(ses);
      if (ctx) lib.SSL_CTX_free(ctx);
    } catch (e) {}
    sys.close(fd);
    return bad(code, text);
  };
  const set_host = (ses, name) => {
    // SSL_CTRL_SET_TLSEXT_HOSTNAME = 55, TLSEXT_NAMETYPE_host_name = 0
    const p = lib.ptr(io_bytes(name + "\0"));
    return lib.SSL_ctrl(ses, 55, 0, p);
  };
  const start_tls = () => {
    let ctx = null;
    let ses = null;
    try {
      ctx = lib.SSL_CTX_new(lib.TLS_client_method());
    } catch (e) {
      ctx = null;
    }
    if (!ctx) {
      return fail(613, "handshake failed", null, null);
    }
    try {
      lib.SSL_CTX_set_verify(ctx, verify ? 1 : 0, null);
      if (verify) lib.SSL_CTX_set_default_verify_paths(ctx);
      const alpn = new Uint8Array([2, 104, 50]);
      lib.SSL_CTX_set_alpn_protos(ctx, lib.ptr(alpn), 3);
      ses = lib.SSL_new(ctx);
      if (!ses) {
        return fail(613, "handshake failed", ctx, null);
      }
      lib.SSL_set_fd(ses, fd);
      set_host(ses, sni);
      if (verify) lib.SSL_set1_host(ses, lib.ptr(io_bytes(sni + "\0")));
    } catch (e) {
      return fail(613, "handshake failed", ctx, ses);
    }
    const step = () => {
      let r = -1;
      try {
        r = lib.SSL_connect(ses);
      } catch (e) {
        return fail(613, "handshake failed", ctx, ses);
      }
      if (r === 1) {
        // ALPN: the client offers only h2 (SSL_CTX_set_alpn_protos
        // above), so OpenSSL itself aborts the handshake when the
        // server selects anything else: r === 1 already proves h2.
        let ok_verify = true;
        try {
          ok_verify = !verify || lib.SSL_get_verify_result(ses) === 0;
        } catch (e) {
          ok_verify = false;
        }
        if (!ok_verify) return fail(614, "verify failed", ctx, ses);
        return io_done({ $: "TlsConn", ssl: ses, ctx: ctx, fd: fd });
      }
      let err = 0;
      try {
        err = lib.SSL_get_error(ses, r);
      } catch (e) {
        return fail(613, "handshake failed", ctx, ses);
      }
      if (err === 2) {
        io_park_on(fd, false, k, step);
        return undefined;
      }
      if (err === 3) {
        io_park_on(fd, true, k, step);
        return undefined;
      }
      let vr = 0;
      try {
        vr = verify ? lib.SSL_get_verify_result(ses) : 0;
      } catch (e) {}
      if (verify && vr !== 0) return fail(614, "verify failed", ctx, ses);
      return fail(613, "handshake failed", ctx, ses);
    };
    return step();
  };
  const again = sys.mac ? 36 : 115;
  const ok = sys.connect(fd, sys.ptr(at), 16) >= 0;
  const code = ok ? 0 : sys.errno();
  if (!ok && code !== again) {
    sys.close(fd);
    return bad(612, "dial failed");
  }
  if (!ok && code === again) {
    const error = () => {
      const v = new Int32Array([0]);
      const l = new Uint32Array([4]);
      return sys.getsockopt(fd, sys.mac ? 0xffff : 1, sys.mac ? 0x1007 : 4,
        sys.ptr(v), sys.ptr(l)) < 0 ? sys.errno() : v[0];
    };
    io_park_on(fd, true, k, () => {
      const ec = error();
      if (ec !== 0) {
        sys.close(fd);
        return bad(612, "dial failed");
      }
      return start_tls();
    });
    return undefined;
  }
  return start_tls();
}

// The libssl symbols this effect needs, dlopened once per process and
// cached in a global slot (the dns_file style). SSL_ctrl replaces the
// set-hostname macro (SSL_set_tlsext_host_name is a macro, not a
// symbol, so dlopen cannot resolve it).
function tls_lib() {
  if (globalThis.BEND_TLS_CONN === undefined) {
    try {
      const ffi = require("bun:ffi");
      const T = { i: "i32", u: "u32", p: "ptr", v: "void" };
      const lib = ffi.dlopen("libssl.so.3", {
        TLS_client_method: { args: [], returns: T.p },
        SSL_CTX_new: { args: [T.p], returns: T.p },
        SSL_CTX_free: { args: [T.p], returns: T.v },
        SSL_CTX_set_verify: { args: [T.p, T.i, T.p], returns: T.v },
        SSL_CTX_set_default_verify_paths: { args: [T.p], returns: T.i },
        SSL_CTX_set_alpn_protos: { args: [T.p, T.p, T.u], returns: T.i },
        SSL_new: { args: [T.p], returns: T.p },
        SSL_free: { args: [T.p], returns: T.v },
        SSL_set_fd: { args: [T.p, T.i], returns: T.i },
        SSL_set1_host: { args: [T.p, T.p], returns: T.i },
        SSL_connect: { args: [T.p], returns: T.i },
        SSL_get_error: { args: [T.p, T.i], returns: T.i },
        SSL_get_verify_result: { args: [T.p], returns: T.i },
        SSL_ctrl: { args: [T.p, T.i, T.i, T.p], returns: T.i },
      });
      globalThis.BEND_TLS_CONN = { ...lib.symbols, ptr: ffi.ptr };
    } catch (e) {
      globalThis.BEND_TLS_CONN = null;
    }
  }
  return globalThis.BEND_TLS_CONN;
}

File: lib/effs/tls_connect.c
// TLS client handshake over a blocking connect on the helper thread.
// The handle answers an SSL session box (TlsConn*), never a raw fd:
// send/recv/close unpack it, so no fd->SSL map is needed and every
// twin stays self-contained (OpenSSL rides dlopen, no link flags).
#include <dlfcn.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <string.h>

// A live session: freed by tls_close only.
#ifndef TLS_CONN_DEFINED
#define TLS_CONN_DEFINED
struct TlsConn {
  void* ssl;
  void* ctx;
  int   fd;
};
#endif

struct TlsDial {
  char* ip;
  u64   ip_len;
  char* sni;
  u64   sni_len;
  u32   port;
  u32   verify;
  u32   alpn;
  u32   code;
  void* conn;
};

// OpenSSL entry points, resolved once per process via dlopen.
static void*          tls_h = NULL;
static pthread_mutex_t tls_gate = PTHREAD_MUTEX_INITIALIZER;
static int            tls_ready = 0;
static void* (*p_TLS_client_method)(void);
static void* (*p_SSL_CTX_new)(void*);
static void  (*p_SSL_CTX_free)(void*);
static void  (*p_SSL_CTX_set_verify)(void*, int, void*);
static int   (*p_SSL_CTX_set_default_verify_paths)(void*);
static int   (*p_SSL_CTX_set_alpn_protos)(void*, const unsigned char*, unsigned);
static void* (*p_SSL_new)(void*);
static int   (*p_SSL_set_fd)(void*, int);
static long  (*p_SSL_ctrl)(void*, int, long, void*);
static int   (*p_SSL_set1_host)(void*, const char*);
static int   (*p_SSL_connect)(void*);
static void  (*p_SSL_get0_alpn_selected)(const void*, const unsigned char**, unsigned*);
static long  (*p_SSL_get_verify_result)(void*);
static int   (*p_SSL_shutdown)(void*);
static void  (*p_SSL_free)(void*);
static int   (*p_SSL_get_error)(void*, int);

static u32 tls_init(void) {
  pthread_mutex_lock(&tls_gate);
  if (tls_ready) {
    pthread_mutex_unlock(&tls_gate);
    return 0;
  }
  tls_h = dlopen("libssl.so.3", RTLD_NOW);
  if (tls_h == NULL) {
    tls_h = dlopen("libssl.so", RTLD_NOW);
  }
  if (tls_h == NULL) {
    pthread_mutex_unlock(&tls_gate);
    return 616;
  }
#define TLS_SYM(n)                                      \
  p_##n = dlsym(tls_h, #n);                             \
  if (p_##n == NULL) {                                  \
    pthread_mutex_unlock(&tls_gate);                    \
    return 616;                                         \
  }
  TLS_SYM(TLS_client_method)
  TLS_SYM(SSL_CTX_new)
  TLS_SYM(SSL_CTX_free)
  TLS_SYM(SSL_CTX_set_verify)
  TLS_SYM(SSL_CTX_set_default_verify_paths)
  TLS_SYM(SSL_CTX_set_alpn_protos)
  TLS_SYM(SSL_new)
  TLS_SYM(SSL_set_fd)
  TLS_SYM(SSL_ctrl)
  TLS_SYM(SSL_set1_host)
  TLS_SYM(SSL_connect)
  TLS_SYM(SSL_get0_alpn_selected)
  TLS_SYM(SSL_get_verify_result)
  TLS_SYM(SSL_shutdown)
  TLS_SYM(SSL_free)
  TLS_SYM(SSL_get_error)
#undef TLS_SYM
  tls_ready = 1;
  pthread_mutex_unlock(&tls_gate);
  return 0;
}

static const char* tls_msg(u32 code) {
  switch (code) {
    case 611: return "bad input";
    case 612: return "dial failed";
    case 613: return "handshake failed";
    case 614: return "verify failed";
    case 615: return "alpn mismatch";
    default:  return "tls error";
  }
}

// Runs on the helper thread: blocking connect + handshake. The loop
// stays free; pack answers on the loop.
static void tls_call(IoWork* w) {
  struct TlsDial* d = (struct TlsDial*)w->data;
  d->code = 0;
  d->conn = NULL;
  u32 q = tls_init();
  if (q != 0) {
    d->code = q;
    return;
  }
  struct sockaddr_in at;
  if (io_nul(d->ip, d->ip_len) || io_nul(d->sni, d->sni_len)
    || io_sys_addr(d->ip, d->port, &at) != 0) {
    d->code = 611;
    return;
  }
  int fd = socket(AF_INET, SOCK_STREAM, 0);
  if (fd < 0) {
    d->code = 612;
    return;
  }
  if (connect(fd, (struct sockaddr*)&at, sizeof(at)) != 0) {
    close(fd);
    d->code = 612;
    return;
  }
  void* ctx = p_SSL_CTX_new(p_TLS_client_method());
  if (ctx == NULL) {
    close(fd);
    d->code = 613;
    return;
  }
  if (d->verify) {
    p_SSL_CTX_set_verify(ctx, 1, NULL);
    p_SSL_CTX_set_default_verify_paths(ctx);
  } else {
    p_SSL_CTX_set_verify(ctx, 0, NULL);
  }
  // ALPN mode: alpn=1 offers h2 and requires it selected (HTTPS);
  // alpn=0 offers nothing and accepts no negotiation (WSS/HTTP/1.1).
  if (d->alpn) {
    const unsigned char alpn[] = {2, 'h', '2'};
    p_SSL_CTX_set_alpn_protos(ctx, alpn, 3);
  }
  void* ssl = p_SSL_new(ctx);
  if (ssl == NULL) {
    p_SSL_CTX_free(ctx);
    close(fd);
    d->code = 613;
    return;
  }
  // SSL_set_tlsext_host_name is a macro: SSL_ctrl(ssl, 55, 0, name)
  // with SSL_CTRL_SET_TLSEXT_HOSTNAME = 55, TLSEXT_NAMETYPE_host_name = 0.
  p_SSL_ctrl(ssl, 55, 0, d->sni);
  if (d->verify) {
    p_SSL_set1_host(ssl, d->sni);
  }
  p_SSL_set_fd(ssl, fd);
  if (p_SSL_connect(ssl) <= 0) {
    long vr = d->verify ? p_SSL_get_verify_result(ssl) : 0;
    p_SSL_shutdown(ssl);
    p_SSL_free(ssl);
    p_SSL_CTX_free(ctx);
    close(fd);
    d->code = (d->verify && vr != 0) ? 614 : 613;
    return;
  }
  if (d->verify && p_SSL_get_verify_result(ssl) != 0) {
    p_SSL_shutdown(ssl);
    p_SSL_free(ssl);
    p_SSL_CTX_free(ctx);
    close(fd);
    d->code = 614;
    return;
  }
  if (d->alpn) {
    const unsigned char* sel = NULL;
    unsigned sel_len = 0;
    p_SSL_get0_alpn_selected(ssl, &sel, &sel_len);
    if (sel_len != 2 || sel == NULL || sel[0] != 'h' || sel[1] != '2') {
      p_SSL_shutdown(ssl);
      p_SSL_free(ssl);
      p_SSL_CTX_free(ctx);
      close(fd);
      d->code = 615;
      return;
    }
  }
  int fl = fcntl(fd, F_GETFL, 0);
  if (fl < 0 || fcntl(fd, F_SETFL, fl | O_NONBLOCK) < 0) {
    p_SSL_shutdown(ssl);
    p_SSL_free(ssl);
    p_SSL_CTX_free(ctx);
    close(fd);
    d->code = 616;
    return;
  }
  struct TlsConn* c = io_mem(malloc(sizeof(*c)));
  c->ssl = ssl;
  c->ctx = ctx;
  c->fd  = fd;
  d->conn = c;
  d->code = 0;
}

static Term tls_pack(Env e, IoWork* w) {
  struct TlsDial* d = (struct TlsDial*)w->data;
  Term r;
  if (d->code == 0) {
    r = io_done(e, io_hand((intptr_t)d->conn));
  } else {
    r = io_fail(e, d->code, tls_msg(d->code));
  }
  free(d->ip);
  free(d->sni);
  free(d);
  return r;
}

Term tls_connect_raw_run(Env e, Term* f, IoWork* w) {
  struct TlsDial* d = io_mem(malloc(sizeof(*d)));
  d->ip     = io_cstr(e, f[0], &d->ip_len);
  d->sni    = io_cstr(e, f[1], &d->sni_len);
  d->port   = (u32)f[2];
  d->verify = (u32)f[3];
  d->alpn   = (u32)f[4];
  d->code   = 0;
  d->conn   = NULL;
  w->data = (char*)d;
  return io_work(w, tls_call, tls_pack);
}

static void __attribute__((constructor)) tls_connect_raw_use(void) {
  io_eff(CID_TLS_CONNECT_RAW, tls_connect_raw_run, 0);
}
